runtime/platform.c

```diff
@@ -0,0 +1,276 @@
+  cond->mutex=0;
+}
+
+
+/* Memory management */
+
+#define Is_power_2(align) \
+  ((align) != 0 && ((align) & ((align) - 1)) == 0)
+
+static uintnat round_up(uintnat size, uintnat align) {
+  CAMLassert(Is_power_2(align));
+  return (size + align - 1) & ~(align - 1);
+}
+
+long caml_sys_pagesize = 0;
+
+uintnat caml_mem_round_up_pages(uintnat size)
+{
+  return round_up(size, caml_sys_pagesize);
+}
+
+#ifdef _WIN32
+#define MAP_FAILED 0
+#endif
+
+void* caml_mem_map(uintnat size, uintnat alignment, int reserve_only)
+{
+  uintnat alloc_sz = caml_mem_round_up_pages(size + alignment);
+  void* mem;
+  uintnat base, aligned_start, aligned_end;
+
+  CAMLassert(Is_power_2(alignment));
+  alignment = caml_mem_round_up_pages(alignment);
+
+  CAMLassert (alloc_sz > size);
+#ifdef _WIN32
+  /* Memory is only reserved at this point. It'll be committed after the
+     trim. */
+  mem = VirtualAlloc(NULL, alloc_sz, MEM_RESERVE, PAGE_NOACCESS);
+#else
+  mem = mmap(0, alloc_sz, reserve_only ? PROT_NONE : (PROT_READ | PROT_WRITE),
+             MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+#endif
+  if (mem == MAP_FAILED) {
+    return 0;
+  }
+
+  /* trim to an aligned region */
+  base = (uintnat)mem;
+  aligned_start = round_up(base, alignment);
+  aligned_end = aligned_start + caml_mem_round_up_pages(size);
+#ifdef _WIN32
+  /* VirtualFree can be used to decommit portions of memory, but it can only
+     release the entire block of memory. For Windows, repeat the call but this
+     time specify the address. */
+  if (!VirtualFree(mem, 0, MEM_RELEASE))
+    printf("The world seems to be upside down\n");
+  mem = VirtualAlloc((void*)aligned_start,
+                     aligned_end - aligned_start + 1,
+                     MEM_RESERVE | (reserve_only ? 0 : MEM_COMMIT),
+                     reserve_only ? PAGE_NOACCESS : PAGE_READWRITE);
+  if (!mem)
+    printf("Trimming failed\n");
+  else if (mem != (void*)aligned_start)
+    printf("Hang on a sec - it's allocated a different block?!\n");
+#else
+  caml_mem_unmap((void*)base, aligned_start - base);
+  caml_mem_unmap((void*)aligned_end, (base + alloc_sz) - aligned_end);
+#endif
+  return (void*)aligned_start;
+}
+
+#ifndef _WIN32
+static void* map_fixed(void* mem, uintnat size, int prot)
+{
+  if (mmap((void*)mem, size, prot,
+           MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED,
+           -1, 0) == MAP_FAILED) {
+    return 0;
+  } else {
+    return mem;
+  }
+}
+#endif
+
+void* caml_mem_commit(void* mem, uintnat size)
+{
+#ifdef _WIN32
+  return VirtualAlloc(mem, size, MEM_COMMIT, PAGE_READWRITE);
+#else
+  void* p = map_fixed(mem, size, PROT_READ | PROT_WRITE);
+  /*
+    FIXME: On Linux, with overcommit, you stand a better
+    chance of getting good error messages in OOM conditions
+    by forcing the kernel to allocate actual memory by touching
+    all the pages. Not sure whether this is a good idea, though.
+
+      if (p) memset(p, 0, size);
```

Comment:
Not sure either.  But surely there are cheaper ways to touch every page than writing zeros to every byte!

Reply:
`MAP_POPULATE` seems promising here and would avoid having to trigger a page-fault on each (4K) page.
Here are some measurements from my system: https://gist.github.com/edwintorok/0a458872c594f98946dc2743df56b9d5#file-results-txt

To summarize:

|method|time|pagefaults|
|-|-|-|
|none|0.0003871 +- 0.0000133 seconds time elapsed  ( +-  3.44% )|49|
|memset|7.9593 +- 0.0167 seconds time elapsed  ( +-  0.21% )|8,388,658|
|touch4k|7.5761 +- 0.0278 seconds time elapsed  ( +-  0.37% )|8,388,656|
|populate|6.07737 +- 0.00877 seconds time elapsed  ( +-  0.14% )|49|

`populate` is faster than `memset` (and avoids doing millions of pagefaults like `memset` would), and has same OOM kill behaviour as `memset` (I get killed close to allocation time, instead of at some random time later when I happen to use the pages) with all 3 methods except none:
```
./map 32 6 memset
[2]    26548 killed     ./map 32 6 memset
./map 32 6 populate
[2]    26578 killed     ./map 32 6 populate
./map 32 6 touch4k
[2]    26968 killed     ./map 32 6 touch4k
./map 32 6 none
```

(note I had to stop `systemd-oomd` on Fedora 35 to do this test, otherwise it'd OOM kill my terminal tab instead of `./map`)

Not doing anything (method `none`) is nearly instant, but you have to pay the cost of page-faults later in the application when you use the pages.

Reply:
It is a good idea to use `MAP_POPULATE` to avoid repeated page faults later on (also in `caml_mem_map` used for allocating pools in the major heap).

But I do not think it is useful for OOM conditions on overcommitting systems. You still get a process killed by the kernel, you cannot handle the error and in fact there is no guarantee that the process killed is the same as the one that called `mmap`. (In your test I suspect that you use a much larger allocation than OCaml does, making it always chosen by the kernel OOM killer; and as you mention there can be user-space OOM killers to deal with.)

The original comment has been there unchanged for a while and I propose to remove it. I do not think there is much to gain in trying to fight the overcommitting behaviour on systems that have it enabled. Not failing immediately is overcommitting working as intended (even if it does not produce any interesting difference in this case), and one should instead focus on reliable ways to handle OOM if there is interest in this topic.

Reply:
Tracking this for post-merge fixing in https://github.com/ocaml/ocaml/issues/10861